// lib-react-components/commonjs/components/tabs/tab.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _analytics_hoc = require('../../containers/analytics_hoc');

var _analytics_hoc2 = _interopRequireDefault(_analytics_hoc);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Tab component
 */
function Tab(props) {
  var _classNames2;

  var className = props.className,
      disabled = props.disabled,
      value = props.value,
      children = props.children,
      selected = props.selected,
      onClick = props.onClick,
      color = props.color,
      colorOn = props.colorOn,
      component = props.component,
      other = _objectWithoutProperties(props, ['className', 'disabled', 'value', 'children', 'selected', 'onClick', 'color', 'colorOn', 'component']);

  /**
   * Tab click handler
   * @param {SytheticEvent} e
   */


  function onClickHandler(e) {
    if (!disabled) {
      onClick(e, value);
    }
  }

  var componentProps = Object.assign({
    'data-component': 'tab',
    'data-selected': selected,
    onClick: onClickHandler,
    className: (0, _classnames2.default)('tab', 'h5', 'break_word', _defineProperty({}, 'text_' + color, !selected), (_classNames2 = {}, _defineProperty(_classNames2, 'text_' + colorOn, selected), _defineProperty(_classNames2, 'stroke_' + colorOn, selected), _classNames2), className),
    disabled: disabled || selected,
    children: _react2.default.createElement(
      'span',
      null,
      children
    )
  }, other);

  if (component) {
    return component(componentProps);
  }

  return _react2.default.createElement('button', _extends({
    type: 'button'
  }, componentProps));
}

Tab.propTypes = {
  /**
   * The css class name of the root element
   */
  className: _propTypes2.default.string,
  /**
   * Disabled the Tab if set to true
   */
  disabled: _propTypes2.default.bool,
  /**
   * If value prop passed to Tabs component, this value prop is also required.
   * It assigns a value to the tab so that it can be selected by the Tabs
   */
  value: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]).isRequired,
  /**
   * The contents of the Tab
   */
  children: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string, _propTypes2.default.node]).isRequired,
  /**
   * Selected the Tab if set to true
   */
  selected: _propTypes2.default.bool,
  /**
   * Callback function that is fired when the Tab clicked
   */
  onClick: _propTypes2.default.func,
  /**
   * Component color from theme
   */
  color: _propTypes2.default.string,
  /**
   * Component checked color from theme
   */
  colorOn: _propTypes2.default.string,
  /**
   * The function component for render custom element
   */
  component: _propTypes2.default.func
};

Tab.defaultProps = {
  className: '',
  disabled: false,
  selected: false,
  onClick: function onClick() {},

  color: 'grey',
  colorOn: 'primary',
  component: undefined
};

exports.default = (0, _analytics_hoc2.default)(Tab, 'onClick');
