// lib-react-components/esm/utils/device_info.js
/**
 * Get window width and height
 */
export function getWindowSize() {
  var _window = window,
      innerWidth = _window.innerWidth,
      innerHeight = _window.innerHeight;


  return {
    width: innerWidth,
    height: innerHeight
  };
}

/**
 * Get window width and height
 */
export function getElementSize(element) {
  var offsetWidth = element.offsetWidth,
      offsetHeight = element.offsetHeight;


  return {
    width: offsetWidth,
    height: offsetHeight
  };
}

/**
 * Get window orietation, depends on rotation angle
 */
export function getWindowOrientation() {
  var _window2 = window,
      orientation = _window2.orientation;


  switch (orientation) {
    case 90 || -90:
      return 'landscape';

    default:
      return 'portrait';
  }
}

/**
 * Get device type and sizes
 */
export function getDeviceInfo() {
  var element = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : window;

  var _ref = element === window ? getWindowSize() : getElementSize(element),
      width = _ref.width,
      height = _ref.height;

  var type = 'desktop';

  if (width <= 1024 && width > 736 && height > 375) {
    type = 'tablet';
  } else if (width <= 736 || width <= 812 && height <= 375) {
    type = 'mobile';
  }

  return {
    width: width,
    height: height,
    type: type,
    orientation: getWindowOrientation()
  };
}
