// lib-react-components/commonjs/components/progress/circular_progress.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = CircularProgress;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var SIZE = 50;

/**
 * CircularProgress component
 */
function CircularProgress(props) {
  var className = props.className,
      color = props.color,
      colorProgress = props.colorProgress,
      size = props.size,
      style = props.style,
      thickness = props.thickness,
      other = _objectWithoutProperties(props, ['className', 'color', 'colorProgress', 'size', 'style', 'thickness']);

  return _react2.default.createElement(
    'div',
    _extends({
      'data-component': 'circular-progress',
      className: (0, _classnames2.default)('circle_progress', className),
      style: _extends({
        width: size,
        height: size
      }, style)
    }, other),
    _react2.default.createElement(
      'svg',
      {
        className: 'circle_progress_svg',
        viewBox: '0 0 ' + SIZE + ' ' + SIZE
      },
      _react2.default.createElement('circle', {
        className: ['circle_progress_stroke_' + color],
        cx: SIZE / 2,
        cy: SIZE / 2,
        r: SIZE / 2 - 5,
        fill: 'none',
        strokeWidth: thickness
      }),
      _react2.default.createElement('circle', {
        className: (0, _classnames2.default)('circle_progress_circle', ['circle_progress_stroke_' + colorProgress]),
        cx: SIZE / 2,
        cy: SIZE / 2,
        r: SIZE / 2 - 5,
        fill: 'none',
        strokeWidth: thickness
      })
    )
  );
}

CircularProgress.propTypes = {
  /**
   * The CSS class name of the root element.
   */
  className: _propTypes2.default.string,
  /**
   * The color of the static circle element.
   */
  color: _propTypes2.default.string,
  /**
   * The color of the progress circle element.
   */
  colorProgress: _propTypes2.default.string,
  /**
   * The size of the circle.
   */
  size: _propTypes2.default.number,
  /**
   * The style of the root element.
   */
  style: _propTypes2.default.oneOfType([_propTypes2.default.object]),
  /**
   * The thickness of the circle.
   */
  thickness: _propTypes2.default.number
};

CircularProgress.defaultProps = {
  className: '',
  color: 'light_grey',
  colorProgress: 'primary',
  size: 24,
  style: {},
  thickness: 4
};
