// lib-react-components/esm/components/progress/circular_progress.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

var SIZE = 50;

/**
 * CircularProgress component
 */
export default function CircularProgress(props) {
  var className = props.className,
      color = props.color,
      colorProgress = props.colorProgress,
      size = props.size,
      style = props.style,
      thickness = props.thickness,
      other = _objectWithoutProperties(props, ['className', 'color', 'colorProgress', 'size', 'style', 'thickness']);

  return React.createElement(
    'div',
    _extends({
      'data-component': 'circular-progress',
      className: classNames('circle_progress', className),
      style: _extends({
        width: size,
        height: size
      }, style)
    }, other),
    React.createElement(
      'svg',
      {
        className: 'circle_progress_svg',
        viewBox: '0 0 ' + SIZE + ' ' + SIZE
      },
      React.createElement('circle', {
        className: ['circle_progress_stroke_' + color],
        cx: SIZE / 2,
        cy: SIZE / 2,
        r: SIZE / 2 - 5,
        fill: 'none',
        strokeWidth: thickness
      }),
      React.createElement('circle', {
        className: classNames('circle_progress_circle', ['circle_progress_stroke_' + colorProgress]),
        cx: SIZE / 2,
        cy: SIZE / 2,
        r: SIZE / 2 - 5,
        fill: 'none',
        strokeWidth: thickness
      })
    )
  );
}

CircularProgress.propTypes = {
  /**
   * The CSS class name of the root element.
   */
  className: PropTypes.string,
  /**
   * The color of the static circle element.
   */
  color: PropTypes.string,
  /**
   * The color of the progress circle element.
   */
  colorProgress: PropTypes.string,
  /**
   * The size of the circle.
   */
  size: PropTypes.number,
  /**
   * The style of the root element.
   */
  style: PropTypes.oneOfType([PropTypes.object]),
  /**
   * The thickness of the circle.
   */
  thickness: PropTypes.number
};

CircularProgress.defaultProps = {
  className: '',
  color: 'light_grey',
  colorProgress: 'primary',
  size: 24,
  style: {},
  thickness: 4
};
