// lib-react-components/esm/components/progress/linear_progress.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * LinearProgress component
 */
export default function LinearProgress(props) {
  var value = props.value,
      color = props.color,
      colorProgress = props.colorProgress,
      className = props.className,
      other = _objectWithoutProperties(props, ['value', 'color', 'colorProgress', 'className']);

  /**
   * @return {ReactElement} markup
   */


  return React.createElement(
    'div',
    _extends({
      'data-component': 'linear-progress',
      className: classNames('linear_progress', 'round_small', ['fill_' + color], className)
    }, other),
    React.createElement('div', {
      className: classNames('linear_progress_value', 'round_small', ['fill_' + colorProgress]),
      style: {
        width: value + '%'
      }
    })
  );
}

/**
 * Verifies value is within the min/max range.
 * @param {object} props
 * @param {string} propName
 * @param {string} componentName
 * @returns {object}
 */
function valueInRangePropType(props, propName, componentName) {
  for (var _len = arguments.length, rest = Array(_len > 3 ? _len - 3 : 0), _key = 3; _key < _len; _key++) {
    rest[_key - 3] = arguments[_key];
  }

  var error = PropTypes.number.apply(PropTypes, [props, propName, componentName].concat(rest));

  if (error !== null) {
    return error;
  }

  var value = props[propName];

  if (value < 0 || value > 100) {
    return new Error(propName + ' should be within the range specified by 0 and 100');
  }

  return null;
}

LinearProgress.propTypes = {
  /**
   * Progress line color from theme
   */
  color: PropTypes.string,
  /**
   * Background line color from theme
   */
  colorProgress: PropTypes.string,
  /**
   * Current progress value
   */
  value: valueInRangePropType,
  /**
   * The CSS class name of the root element
   */
  className: PropTypes.string
};

LinearProgress.defaultProps = {
  color: 'light_grey',
  colorProgress: 'primary',
  value: 0,
  className: ''
};
