// lib-react-components/esm/components/typography/index.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export var textTypePropType = PropTypes.oneOfType([PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'b1', 'b2', 'b3', 'c1', '']), PropTypes.string]);

/**
 * Typography component
 */
export default function Typography(props, context) {
  var children = props.children,
      propsType = props.type,
      color = props.color,
      align = props.align,
      tabletType = props.tabletType,
      mobileType = props.mobileType,
      className = props.className,
      other = _objectWithoutProperties(props, ['children', 'type', 'color', 'align', 'tabletType', 'mobileType', 'className']);

  var device = context.device;

  var type = propsType || 'c1';

  if (device) {
    if (mobileType && device.type === 'mobile') {
      type = mobileType;
    }

    if (tabletType && device.type === 'tablet') {
      type = tabletType;
    }
  }

  var basicClassNames = classNames('break_word', ['text_' + color], _defineProperty({}, 'text_' + align, align !== 'auto'), className);
  var Component = /h[1-6]/.test(type) ? type : 'p';

  return React.createElement(
    Component,
    _extends({}, other, {
      'data-component': 'typography',
      'data-classnamemobile': mobileType ? classNames(basicClassNames, mobileType) : null,
      'data-classnametablet': tabletType ? classNames(basicClassNames, tabletType) : null,
      'data-classnamedesktop': mobileType || tabletType ? basicClassNames : null,
      className: classNames(basicClassNames, type)
    }),
    children
  );
}

Typography.propTypes = {
  /**
   * This is what will be displayed inside the button
   */
  children: PropTypes.node.isRequired,
  /**
   * Typography type
   */
  type: textTypePropType,
  /**
   * Typography type for tablet
   */
  tabletType: textTypePropType,
  /**
   * Typography type for mobile
   */
  mobileType: textTypePropType,
  /**
   * Component color from theme
   */
  color: PropTypes.string,
  /**
   * Component content aligment
   */
  align: PropTypes.oneOf(['left', 'center', 'right', 'auto']),
  /**
   * The CSS class name of the root element
   */
  className: PropTypes.string
};

Typography.contextTypes = {
  device: PropTypes.object
};

Typography.defaultProps = {
  type: 'b1',
  color: 'black',
  align: 'left',
  className: '',
  tabletType: '',
  mobileType: ''
};
