// lib-react-components/esm/components/progressive_image/index.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react'; // eslint-disable-line
import PropTypes from 'prop-types';

/**
 * ProgressiveImage component
 */

var ProgressiveImage = function (_PureComponent) {
  _inherits(ProgressiveImage, _PureComponent);

  function ProgressiveImage() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ProgressiveImage);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ProgressiveImage.__proto__ || Object.getPrototypeOf(ProgressiveImage)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      image: _this.props.placeholder,
      loading: true
    }, _this._onLoad = function () {
      var src = _this.props.src;


      _this.setState({
        image: src,
        loading: false
      });
    }, _this._onError = function (e) {
      var onError = _this.props.onError;


      if (onError) {
        onError(e);
      }
    }, _this._loadImage = function (src) {
      if (_this.image) {
        _this.image.onload = null;
        _this.image.onerror = null;
      }

      var image = new Image();

      _this.image = image;
      image.onload = _this._onLoad;
      image.onerror = _this._onError;
      image.src = src;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ProgressiveImage, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var src = this.props.src;


      this._loadImage(src);
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var _this2 = this;

      var src = this.props.src;
      var nextSrc = nextProps.src,
          placeholder = nextProps.placeholder;


      if (nextSrc !== src) {
        this.setState({
          image: placeholder,
          loading: true
        }, function () {
          _this2._loadImage(nextSrc);
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.image) {
        this.image.onload = null;
        this.image.onerror = null;
      }
    }

    /**
     * On load handler
     */


    /**
     * On error load handler
     * @param {SytheticEvent} e
     */


    /**
     * Load image handler
     * @param {string} src
     */

  }, {
    key: 'render',


    /**
     * render
     * @return {ReactElement} markup
     */
    value: function render() {
      var _state = this.state,
          image = _state.image,
          loading = _state.loading;
      var children = this.props.children;


      if (!children || typeof children !== 'function') {
        throw new Error('ProgressiveImage requires a function as its only child');
      }

      return children(image, loading);
    }
  }]);

  return ProgressiveImage;
}(PureComponent);

ProgressiveImage.propTypes = {
  children: PropTypes.func.isRequired,
  /**
   * Fired when an error occurs when uploading a picture
   */
  onError: PropTypes.func,
  /**
   * Path to preloaded image
   */
  placeholder: PropTypes.string,
  /**
   * Path to main image
   */
  src: PropTypes.string
};
ProgressiveImage.defaultProps = {
  onError: function onError() {},

  placeholder: '',
  src: ''
};
export default ProgressiveImage;
