// lib-react-components/esm/containers/analytics_hoc/index.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';

/**
 * AnalyticsHOC
 */
export default function withAnalytics(Component) {
  var firingEventProp = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'onClick';

  var AnalyticsWrapper = function AnalyticsWrapper(_ref, context) {
    var forwardRef = _ref.forwardRef,
        gaEventName = _ref.gaEventName,
        props = _objectWithoutProperties(_ref, ['forwardRef', 'gaEventName']);

    var oldProp = props[firingEventProp];
    var fireEvent = function fireEvent() {
      if (context.gaFireEvent && gaEventName) {
        context.gaFireEvent(gaEventName);
      }

      if (typeof oldProp === 'function') {
        oldProp.apply(undefined, arguments);
      }
    };

    return React.createElement(Component, _extends({}, props, _defineProperty({}, firingEventProp, fireEvent), {
      ref: forwardRef
    }));
  };

  var WrappedComponent = React.forwardRef(function (props, ref) {
    return React.createElement(AnalyticsWrapper, _extends({}, props, { forwardRef: ref }));
  });

  WrappedComponent.name = Component.name;
  AnalyticsWrapper.contextTypes = {
    gaFireEvent: PropTypes.func
  };
  AnalyticsWrapper.propTypes = {
    forwardRef: PropTypes.oneOfType([PropTypes.func, PropTypes.shape({ current: PropTypes.any })]),
    gaEventName: PropTypes.string
  };
  AnalyticsWrapper.defaultProps = {
    forwardRef: null,
    gaEventName: ''
  };

  return WrappedComponent;
}
