// lib-react-components/commonjs/components/highlight_code/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _prismjs = require('prismjs');

var _prismjs2 = _interopRequireDefault(_prismjs);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

require('prismjs/components/prism-jsx');

require('prismjs/components/prism-bash');

require('prismjs/components/prism-glsl');

require('prismjs/components/prism-json');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Highlight component
 */
var HighlightCode = function (_PureComponent) {
  _inherits(HighlightCode, _PureComponent);

  function HighlightCode() {
    _classCallCheck(this, HighlightCode);

    return _possibleConstructorReturn(this, (HighlightCode.__proto__ || Object.getPrototypeOf(HighlightCode)).apply(this, arguments));
  }

  _createClass(HighlightCode, [{
    key: 'componentWillMount',


    /**
     * Fix Prism removes <br/> on Prism.highlightAll()
     * https://github.com/PrismJS/prism/issues/832
     */
    value: function componentWillMount() {
      _prismjs2.default.hooks.add('before-highlight', function (env) {
        env.code = env.element.innerText;
      });
    }
  }, {
    key: '_getPlainCode',
    value: function _getPlainCode() {
      var children = this.props.children;

      if (Array.isArray(children)) {
        return children.join(' ');
      }

      return String(children);
    }

    /**
     * Hightlight element
     */

  }, {
    key: '_hightlight',
    value: function _hightlight() {
      var lang = this.props.lang;

      var grammar = _prismjs2.default.languages[lang];
      var _html = this._getPlainCode();

      if (grammar) {
        try {
          _html = _prismjs2.default.highlight(_html, grammar, lang);
        } catch (err) {
          console.warn('React Components Highlighter: There is an error highlighting lang ' + lang, err);
        }
      } else {
        _html = _html.replace(/[<>]/g, function (s) {
          return '&#' + s.charCodeAt(0) + ';';
        });
      }

      // Treat iOS whie-space: pre; behavior
      return _html.replace(/\n/g, '<br>').replace(/<br>[ ]+/g, function (str) {
        return '<br>' + '&nbsp;'.repeat(str.length - 5);
      }).replace(/^[ ]+/g, function (str) {
        return '&nbsp;'.repeat(str.length);
      });
    }

    /**
     * render
     * @return {ReactElement} markup
     */

  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          children = _props.children,
          lang = _props.lang,
          other = _objectWithoutProperties(_props, ['children', 'lang']);

      return _react2.default.createElement(
        'pre',
        _extends({
          'data-component': 'highlight_code'
        }, other, {
          className: (0, _classnames2.default)(other.className, 'language-' + lang)
        }),
        _react2.default.createElement('code', {
          style: {
            whiteSpace: 'nowrap'
          },
          ref: function ref(node) {
            _this2.rootNode = node;
          },
          className: 'language-' + lang,
          dangerouslySetInnerHTML: { __html: this._hightlight() }
        })
      );
    }
  }]);

  return HighlightCode;
}(_react.PureComponent);

HighlightCode.propTypes = {
  /**
   * This is what will be displayed inside the component
   */
  children: _propTypes2.default.node.isRequired,
  /**
   * Supported highlighting languages
   */
  lang: _propTypes2.default.oneOf(['css', 'html', 'javascript', 'js', 'jsx', 'markup', 'svg', 'xml', 'bash', 'glsl', 'none', 'json'])
};
HighlightCode.defaultProps = {
  lang: 'js' };
exports.default = HighlightCode;
