// lib-react-components/esm/components/select/select_item.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * SelectItem component
 */
export default function SelectItem(props) {
  var _classNames;

  var children = props.children,
      value = props.value,
      selected = props.selected,
      disabled = props.disabled,
      hasFocus = props.hasFocus,
      className = props.className,
      textColor = props.textColor,
      colorFocus = props.colorFocus,
      size = props.size,
      other = _objectWithoutProperties(props, ['children', 'value', 'selected', 'disabled', 'hasFocus', 'className', 'textColor', 'colorFocus', 'size']);

  return React.createElement(
    'div',
    _extends({
      'data-component': 'select_item',
      'data-disabled': disabled,
      'data-selected': selected,
      'data-has-focus': hasFocus,
      'data-value': value,
      className: classNames('select_item', 'truncate_text', ['select_item_' + size], (_classNames = {}, _defineProperty(_classNames, 'select_item_focus_' + colorFocus, hasFocus), _defineProperty(_classNames, 'text_' + textColor, !selected), _classNames), _defineProperty({}, 'text_' + colorFocus, selected), className)
    }, other),
    children
  );
}

SelectItem.propTypes = {
  children: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.node]).isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  selected: PropTypes.bool,
  disabled: PropTypes.bool,
  hasFocus: PropTypes.bool,
  className: PropTypes.string,
  /**
   * Component text color from theme
   */
  textColor: PropTypes.string,
  /**
   * Component focus color from theme
   */
  colorFocus: PropTypes.string,
  /**
   * Component size
   */
  size: PropTypes.oneOf(['medium', 'large'])
};

SelectItem.defaultProps = {
  selected: false,
  disabled: false,
  hasFocus: false,
  className: '',
  textColor: 'black',
  colorFocus: 'primary',
  size: 'medium'
};
