// lib-react-components/esm/components/content_loader/index.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import uuid from '../../utils/uuid';

/**
 * ContentLoader component
 */

var ContentLoader = function (_PureComponent) {
  _inherits(ContentLoader, _PureComponent);

  function ContentLoader() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ContentLoader);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ContentLoader.__proto__ || Object.getPrototypeOf(ContentLoader)).call.apply(_ref, [this].concat(args))), _this), _this.idClip = uuid(), _this.idGradient = uuid(), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ContentLoader, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          color = _props.color,
          duration = _props.duration,
          className = _props.className,
          height = _props.height,
          width = _props.width,
          children = _props.children,
          other = _objectWithoutProperties(_props, ['color', 'duration', 'className', 'height', 'width', 'children']);

      return React.createElement(
        'svg',
        _extends({
          'data-component': 'content-loader',
          viewBox: '0 0 ' + width + ' ' + height,
          className: classNames('content_loader', className)
        }, other),
        React.createElement('rect', {
          style: { fill: 'url(#' + this.idGradient + ')' },
          clipPath: 'url(#' + this.idClip + ')',
          x: '0',
          y: '0',
          width: width,
          height: height
        }),
        React.createElement(
          'defs',
          null,
          React.createElement(
            'clipPath',
            { id: this.idClip },
            children
          ),
          React.createElement(
            'linearGradient',
            { id: this.idGradient },
            React.createElement('stop', { offset: '0%', className: 'content_loader_color_' + color }),
            React.createElement('stop', { offset: '50%', className: 'content_loader_color_' + color }),
            React.createElement('stop', { offset: '100%', className: 'content_loader_color_' + color })
          )
        )
      );
    }
  }]);

  return ContentLoader;
}(PureComponent);

ContentLoader.propTypes = {
  /**
   * Color from theme
   */
  color: PropTypes.string,
  /**
   * Animation duration
   */
  duration: PropTypes.number,
  /**
   * The CSS class name of the root element
   */
  className: PropTypes.string,
  /**
   * Height of component in px
   */
  height: PropTypes.number,
  /**
   * Width of component in px
   */
  width: PropTypes.number,
  /**
   * The content of the component
   */
  children: PropTypes.node.isRequired
};
ContentLoader.defaultProps = {
  color: 'grey',
  duration: 2,
  className: '',
  height: 300,
  width: 300
};
export default ContentLoader;
